core/types/transaction.go

```diff
@@ -49,75 +51,57 @@ type TransactionType byte
+	CreateValidator
+	EditValidator
 	Delegate
 	Undelegate
 	CollectRewards
+	InvalidTx
 )
 
-// StakingTypeMap is the map from staking type to transactionType
-var StakingTypeMap = map[staking.Directive]TransactionType{staking.DirectiveCreateValidator: StakeCreateVal,
-	staking.DirectiveEditValidator: StakeEditVal, staking.DirectiveDelegate: Delegate,
-	staking.DirectiveUndelegate: Undelegate, staking.DirectiveCollectRewards: CollectRewards}
-
-// Transaction struct.
-type Transaction struct {
-	data txdata
-	// caches
-	hash atomic.Value
-	size atomic.Value
-	from atomic.Value
-}
-
-// RPCTransactionError ..
-type RPCTransactionError struct {
-	TxHashID             string `json:"tx-hash-id"`
-	TimestampOfRejection int64  `json:"time-at-rejection"`
-	ErrMessage           string `json:"error-message"`
-}
-
-// NewRPCTransactionError ...
-func NewRPCTransactionError(hash common.Hash, err error) RPCTransactionError {
-	return RPCTransactionError{
-		TxHashID:             hash.Hex(),
-		TimestampOfRejection: time.Now().Unix(),
-		ErrMessage:           err.Error(),
-	}
-}
-
-//String print mode string
+// String print mode string
 func (txType TransactionType) String() string {
-	if txType == SameShardTx {
+	switch txType {
+	case SameShardTx:
 		return "SameShardTx"
-	} else if txType == SubtractionOnly {
+	case SubtractionOnly:
 		return "SubtractionOnly"
-	} else if txType == InvalidTx {
-		return "InvalidTx"
-	} else if txType == StakeCreateVal {
-		return "StakeNewValidator"
-	} else if txType == StakeEditVal {
-		return "StakeEditValidator"
-	} else if txType == Delegate {
+	case Contract:
+		return "ContractCreation"
+	case CreateValidator:
+		return "CreateValidator"
+	case EditValidator:
+		return "EditValidator"
+	case Delegate:
 		return "Delegate"
-	} else if txType == Undelegate {
+	case Undelegate:
 		return "Undelegate"
-	} else if txType == CollectRewards {
+	case CollectRewards:
 		return "CollectRewards"
+	case InvalidTx:
+		return "InvalidTx"
+	default:
+		return "Unknown"
 	}
-	return "Unknown"
+}
+
+// Transaction struct.
+type Transaction struct {
+	data txdata
+	// caches
+	hash     atomic.Value
+	size     atomic.Value
+	from     atomic.Value
+	blockNum *big.Int
 }
 
 type txdata struct {
 	AccountNonce uint64          `json:"nonce"      gencodec:"required"`
 	Price        *big.Int        `json:"gasPrice"   gencodec:"required"`
 	GasLimit     uint64          `json:"gas"        gencodec:"required"`
-	ShardID      uint32          `json:"shardID"    gencodec:"required"`
-	ToShardID    uint32          `json:"toShardID"  gencodec:"required"`
-	Recipient    *common.Address `json:"to"         rlp:"nil"` // nil means contract creation
-	Amount       *big.Int        `json:"value"      gencodec:"required"`
-	Payload      []byte          `json:"input"      gencodec:"required"`
+	TxType       TransactionType `json:"txType"`
+	Message      interface{}     `json:"message"`
```

Comment:
I thought we agreed to not take this intrusive route? This will need all partners/users to upgrade their clients.